LibraBFT/Impl/Consensus/RoundManager.agda

```diff
@@ -95,6 +98,106 @@ addVoteM now vote = do
       _ →
         pure unit
 
+  processVote : Instant → VoteMsg → LBFT Unit
+  processVote now msg = pure unit
+
+  ------------------------------------------------------------------------------
+  syncUpM : Instant → SyncInfo → Author → LBFT (ErrLog ⊎ Unit)
+  ensureRoundAndSyncUpM : Instant → Round → SyncInfo → Author → Bool →
+                          LBFT (ErrLog ⊎ Bool)
+  processProposalM : Block → LBFT Unit
+  executeAndVoteM : Block → LBFT (ErrLog ⊎ VoteWithMeta)
+
+  -- external entry point
+  -- TODO-2: The sync info that the peer requests if it discovers that its round
+  -- state is behind the sender's should be sent as an additional argument, for now.
+
+  -- This is broken up into smaller pieces in order to aid in the verification effort.
+  processProposalMsgM-check₁ : Instant → ProposalMsg → Author → LBFT Unit
+  processProposalMsgM-check₁-cont : Instant → ProposalMsg → Author → ErrLog ⊎ Bool → LBFT Unit
+
+  processProposalMsgM : Instant → {- Author → -} ProposalMsg → LBFT Unit
+  processProposalMsgM now {- from -} pm
+     with pm ^∙ pmProposer
+  ...| nothing = pure unit -- errorExit "ProposalMsg does not have an author"
```

Comment:
That is a good question. For `processProposalMsgM`, I have been using the output (i.e., whether it has any or not) as an indicator of whether it succeeded for failed. If that is true for the other top-level functions of type `LBFT Unit`, then perhaps this suffices?

Reply:
In Haskell:

```
data Output a
  = BroadcastEpochChangeProof LogControl EpochChangeProof     NodesInOrder
  | BroadcastProposal         LogControl (ProposalMsg a)      NodesInOrder
  | BroadcastSyncInfo         LogControl SyncInfo             NodesInOrder
  | BroadcastVote             LogControl VoteMsg              NodesInOrder
  | LogErr                               (ErrLog a)
  | LogInfo                   LogControl (InfoLog a)
  | ReconfigLocalEpochChange  LogControl ReconfigEventEpochChange
  | SendBRP                   LogControl Author               (BlockRetrievalResponse a)
  | SendVote                  LogControl VoteMsg              NodesInOrder
  | SetTimeout                LogControl Duration             Round
  deriving (Eq, Show)
```
so the errors DO show up in the output.  All the "top level" functions in `RoundManager` return `LBFT Unit` and are expected to catch and report errors from below.

So I do think it suffices.

Do we want real error types rather than the existing fake unit?

Reply:
What this means, is that all the places like:

```
processProposalMsgM : Instant → {- Author → -} ProposalMsg → LBFT Unit
processProposalMsgM now {- from -} pm =
  case pm ^∙ pmProposer of λ where
    nothing        → pure unit -- log: info: proposal with no author
    (just pAuthor) →
      ensureRoundAndSyncUpM now (pm ^∙ pmProposal ∙ bRound) (pm ^∙ pmSyncInfo)
                            pAuthor true >>= λ where
      (inj₁ _)     → pure unit -- log: error: <propagate error>
      (inj₂ true)  → processProposalM (pm ^∙ pmProposal)
      (inj₂ false) → do
        currentRound ← use (lRoundState ∙ rsCurrentRound)
        pure unit -- log: info: dropping proposal for old round
```
where we are returning `pure unit` should do like Haskell and log a specific error into `Output`

Reply:
This is a discussion I would want @mark-moir involved in. My guess is that we would definitely want this eventually, but I do not its absence being being an immediate impediment to the proof effort.

Reply:
Thanks guys, this is an important discussion.  If we construct proofs with the assumption (accurate for our partial model for now) that, for example, `processProposalMsg` produces either no `Output`s or exactly one `Output` which is a `SendVote`, these proofs would break when we introduce more `Output` types later.

On the other hand, we don't want to spend time now accurately modeling logged messages and errors, as we're more interested in correctness in the short term, rather than proving that certain circumstances result in certain error messages, etc.

So, my feeling is that we should make the model accurately reflect the implementation in terms of what *kinds* of outputs it sends, but *not* bother for now with details like actual log/error messages.

As mentioned before, `unit` shows up in our initial proofs for various things, and I'd much see something like `fakeError` in proofs than `unit`.  So maybe we could define:
```
FakeError : Set
postulate
  fakeError : FakeError
```
and have an `Output` constructor like `LogErr FakeError`?

That way the proofs would adapt better when we do refine the model to log actual errors, and we wouldn't fall into the trap mentioned above.

I think we should not hold up this PR for this issue, but we should resolve the discussion and implement what we decide soon, before spending time developing more proofs that will break when we refine the model.

Reply:
FYI, I have done something along these lines in a pull request I have just sent (#47).